### PR TITLE
First class clang-tidy support.

### PR DESCRIPTION
**Pull request by Hoikas**
This adds first class support for clang-tidy for Makefile, Ninja, and Visual Studio generators. It is loosely based on the implementation found in DIRTSAND. ~~Currently, only the modernize-use-nullptr check is enabled; modernize-use-override is still complaining about some destructors, which is outside the scope of this PR.~~

To use clang-tidy:
- If using a Visual Studio **solution**, click `Analyze` > `Run Code Analsys` > `On Solution` - this will run clang-tidy now instead of the Microsoft code analysis.
- In other generators, including Visual Studio CMake, build the `tidy` target. Any checks with fixers can be applied by building the `fix` target.

I had to clean up the SIMD/CPUID stuff a bit for clang-tidy+MSVC to not blow up on the CoreLib and plPipeline targets. As a side effect, SIMD is now used on GCC and Clang, where it evidently was not before.  Due to the way GCC does SIMD intrinsics, each SIMD instruction set must be limited to its own file for cross-platform code; otherwise, GCC may emit instructions that the end user's CPU does not implement. (NOTE: SIMD stuff was split off to #862)

Potentially outside the scope of this PR, but relevant: which checks should we be using? Many of them look useful (eg bugprone-unused-raii) but others are actually legitimate (anti-)patterns we use (eg modernize-avoid-c-arrays).

### CODE REVIEW

**Comment by zrax** [comment]
> modernize-use-override is still complaining about some destructors, which is outside the scope of this PR.

You can set `IgnoreDestructors` to, well, ignore the destructors:  https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-override.html

**Comment by Hoikas** [comment]
I've got the following .clang-tidy file locally, but it still seems to be complaining about destructors.
```
Checks: '-*,modernize-use-override,modernize-use-nullptr'
CheckOptions: [
    {
        key: modernize-use-override.IgnoreDestructors,
        value: 1
    }
]
HeaderFilterRegex: '.*'
```

**Comment by zrax** [comment]
Hmm, I don't know how it's working for you, but for me (using Ninja) the compiler seems to think exceptions aren't enabled, so everything barfs...

File: cmake/Sanitizers.cmake [hunk 1]
```diff
@@ -0,0 +1,67 @@
+find_program(CLANG_APPLY_EXE NAMES "clang-apply-replacements" DOC "Path to clang-apply-replacements executable")
+find_program(CLANG_TIDY_EXE NAMES "clang-tidy" DOC "Path to clang-tidy executable")
+
+cmake_dependent_option(USE_CLANG_TIDY "Allow the codebase to use the clang-tidy sanitizer" ON "CLANG_TIDY_EXE" OFF)
+
+if(CMAKE_GENERATOR MATCHES "[Mm][Aa][Kk][Ee][Ff][Ii][Ll][Ee]|[Nn][Ii][Nn][Jj][Aa]")
```
Comment (zrax) [inline cmake/Sanitizers.cmake:6]: I think this is set to a constant string, so the silliness around case sensitivity is unnecessary...

**Comment by Hoikas** [comment]
Yikes. Is this on Windows? Ninja is working well on Linux for me.

File: cmake/CompilerChecks.cmake [hunk 2]
```diff
@@ -23,11 +23,59 @@ if(NOT DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION)
...
+    cmake_parse_arguments(_passf "" "SOURCE_GROUP" "${_INSTRUCTION_SETS}" ${ARGN})
+
+    # Hack: if we ever bump to CMake 3.17, use ZIP_LISTS.
+    list(LENGTH _INSTRUCTION_SETS _LENGTH)
+    math(EXPR _LENGTH "${_LENGTH} - 1")
+
+    foreach(i RANGE ${_LENGTH})
+        # foreach(i IN ZIP_LISTS _INSTRUCTION_SETS _GCC_ARGS)
+        list(GET _INSTRUCTION_SETS ${i} i_0)
+        list(GET _GCC_ARGS ${i} i_1)
```
Comment (zrax) [inline cmake/CompilerChecks.cmake:58]: Even if you are using ZIP_LISTS, I would prefer to see reasonable variable names, rather than `${i_0}` and `${i_1}`

**Comment by zrax** [comment]
Windows (Ninja + MSVC)

**Comment by Hoikas** [comment]
Hmm... It was "working" with MSVC+CMake last night, but I haven't tested it lately. I'll investigate tomorrow 😄

File: cmake/CompilerChecks.cmake (hunk 2 above)
Comment (Hoikas) [reply cmake/CompilerChecks.cmake:58]: Sadly, we aren't using `ZIP_LISTS`, so reasonable variable names are definitely an option.

File: cmake/Sanitizers.cmake (hunk 1 above)
Comment (Hoikas) [reply cmake/Sanitizers.cmake:6]: I aM rEmInDeD oF wHeN iT wAs KeWl To CaPiTaLiZe LiKe tHiS. :trollface:

**Comment by Hoikas** [comment]
OK, Ninja+MSVC should be good now.

**Comment by zrax** [comment]
I'll try it later when I get the chance...  But I like the option of having `tidy` be a separate build target rather than a normal part of compilation anyway :)

File: cmake/PlasmaTargets.cmake [hunk 3]
```diff
@@ -85,11 +85,12 @@ function(plasma_library TARGET)
...
 
-    #TODO (#854): plasma_sanitize_target(${TARGET})
+    plasma_sanitize_target(${TARGET})
 endfunction()
 
 function(plasma_test TARGET)
     plasma_executable(${TARGET} ${ARGN})
     add_test(NAME ${TARGET} COMMAND ${TARGET})
     add_dependencies(check ${TARGET})
+    plasma_sanitize_target(${TARGET})
```
Comment (zrax) [inline cmake/PlasmaTargets.cmake:95]: `plasma_executable()` already calls `plasma_sanitize_target()`

Comment (Hoikas) [reply cmake/PlasmaTargets.cmake:95]: Whoops!

**Comment by Hoikas** [comment]
I fixed some issues where clang-tidy was trying to tidy bitmaps and Windows resource files, hosing the whole process.

**Comment by Deledrius** [comment]
> 
> 
> I fixed some issues where clang-tidy was trying to tidy bitmaps and Windows resource files, hosing the whole process.

Silly clang, you can't tidy those.

**Comment by Hoikas** [comment]
`USE_CLANG_TIDY` now defaults to `OFF` when using the Unix Makefiles or Ninja generator. Setting `USE_CLANG_TIDY=ON` will disable build optimizations (PCH, Unity) on these generators. This is due to clang-tidy only knowing about, for example, plNetMsgScreener.cpp but not unity_{n}.cxx while the inverse is true for the compile commands JSON file. This results in fascinating but irrelevant errors.

File: cmake/hsConfig.h.cmake [hunk 4]
```diff
@@ -45,6 +45,14 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
...
 /* Compiler settings */
 #cmakedefine HAVE_CPUID
+#cmakedefine HS_AVX2
+#cmakedefine HS_AVX
+#cmakedefine HS_SSE42
+#cmakedefine HS_SSE41
+#cmakedefine HS_SSE4
+#cmakedefine HS_SSE3
+#cmakedefine HS_SSE2
+#cmakedefine HS_SSE1
```
Comment (zrax) [inline cmake/hsConfig.h.cmake:55]: I realize these are the names we used before, but perhaps `HAVE_<instruction set>` would be simpler / more readable?

File: cmake/CompilerChecks.cmake [hunk 5]
```diff
@@ -23,11 +23,67 @@ if(NOT DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION)
     endif()
 endif()
 
-# Check for CPUID headers
-try_compile(HAVE_CPUID ${PROJECT_BINARY_DIR}
-            ${PROJECT_SOURCE_DIR}/cmake/check_cpuid.cpp)
```
Comment (zrax) [inline cmake/CompilerChecks.cmake:28]: The big advantage of the check file is that it checked in exactly the same way as the use in hsCpuID.h...  By pulling this into CMake, the check is no longer the same, which may result in odd errors when someone uses different compilers or tools :/

File: cmake/Sanitizers.cmake [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            )
+        endif()
+
+        if(USE_SANITIZE_TARGETS)
+            get_target_property(_TARGET_SOURCES ${TARGET} SOURCES)
+            get_target_property(_TARGET_SOURCE_DIR ${TARGET} SOURCE_DIR)
+            foreach(_SOURCE_FILE IN LISTS _TARGET_SOURCES)
+                # Prevent CMake from stoopidly feeding Python files and bitmaps to clang-tidy.
+                # Yes, CMake actually does try to do that...
+                if(_SOURCE_FILE MATCHES [[\.c$|\.cpp$|\.cxx$|\.cc$]])
```
Comment (zrax) [inline cmake/Sanitizers.cmake:77]: No headers?

File: cmake/Sanitizers.cmake [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    set(ALLOW_BUILD_OPTIMIZATIONS TRUE)
+endif()
+
+cmake_dependent_option(
+    USE_CLANG_TIDY
+    "Allow the codebase to use the clang-tidy sanitizer"
+    ${ALLOW_BUILD_OPTIMIZATIONS} # Default such that we don't clobber PlasmaTargets.
+    "CLANG_TIDY_EXE"
+    OFF
+)
```
Comment (zrax) [inline cmake/Sanitizers.cmake:27]: Circular dependency:  `USE_CLANG_TIDY` depends on `ALLOW_BUILD_OPTIMIZATIONS`, which is set based on `[...] AND USE_CLANG_TIDY`

File: cmake/Sanitizers.cmake (hunk 6 above)
Comment (Hoikas) [reply cmake/Sanitizers.cmake:77]: We can only tidy files listed in the compile commands JSON. I think clang-tidy smart enough to discard headers, knowing that they'll be checked on include. It isn't smart enough to filter out python files, resource files, or bitmaps though.

File: cmake/CompilerChecks.cmake (hunk 5 above)
Comment (Hoikas) [reply cmake/CompilerChecks.cmake:28]: I could pull the cpuid test back out into its own file, but I think the various simd header checks are good like this.

**Comment by zrax** [comment]
Clang-tidy can't currently handle the Max headers, which breaks the global 'tidy' target -- perhaps we should add an option to skip creating a tidy target for certain libs?

**Comment by Hoikas** [comment]
Or we could add a clang-tidy file that skips the max headers? Max includes should be jailed to MaxAPI.h, so skipping that file should be sufficient... But, now that I'm thinking about it, I thought the global clang-tidy file should prevent any header outside Sources/ from being processed.

File: cmake/CompilerChecks.cmake (hunk 5 above)
Comment (zrax) [reply cmake/CompilerChecks.cmake:28]: Yeah, the SIMD tests should definitely be kept in CMake; I was just looking specifically at the cpuid check, which is more convoluted.

**Comment by zrax** [comment]
The problem isn't that clang-tidy needs to *tidy* the max headers, just that it needs to *parse* them in order to process the rest of the source file...  It's the parsing that clang-tidy has trouble with

File: cmake/Sanitizers.cmake (hunk 6 above)
Comment (zrax) [reply cmake/Sanitizers.cmake:77]: Very odd, but it raises an interesting question:  Why are those files listed as target sources to begin with?  Can't they be added to the source group without being a target source?

Comment (Hoikas) [reply cmake/Sanitizers.cmake:77]: Unfortunately not. Omitting them from the target prevents them from being listed in the IDE at all. Also, you want *.rc files to be compiled into the target by MSVC.

**Comment by Hoikas** [comment]
Oh, right. Yeah, at bare minimum we could have a `NO_TIDY` (or should that be `NO_SANITIZE`?) option on all of our target functions. Maybe with some further investigation, we could find which, if any, Max headers clang can parse.

File: cmake/CompilerChecks.cmake (hunk 5 above)
Comment (Hoikas) [reply cmake/CompilerChecks.cmake:28]: Note for posterity, the SIMD stuff was split off to #862.

**Comment by zrax** [comment]
I'd vote for `NO_SANITIZE` for consistency with `plasma_sanitize_target`

**Comment by Hoikas** [comment]
Ok, I turned off clang-tidy for the Max targets. I saw the checks were firing on some Plasma code, but the Max headers are just too broken for clang to parse... At least with Max 2012.

**Review by zrax — COMMENTED** [review]
It seems gtest headers are caught, since they are under **Sources**/Tests/gtest/, but I'm fine with leaving them if they're tricky to exclude (they're not that noisy).

I also had to make an additional change to make clang happy:
```diff
--- a/Sources/Plasma/PubUtilLib/plPipeline/DX/plDXPipeline.cpp
+++ b/Sources/Plasma/PubUtilLib/plPipeline/DX/plDXPipeline.cpp
@@ -11576,7 +11576,7 @@ void plDXPipeline::IRenderShadowsOntoSpan(const plRenderPrimFunc& render, const
                 if( selfShadowNow )
                 {
                     plConst(float) kMaxSelfPower = 0.3f;
-                    float power = fShadows[i]->fPower > kMaxSelfPower ? kMaxSelfPower : fShadows[i]->fPower;
+                    float power = fShadows[i]->fPower > kMaxSelfPower ? (float)kMaxSelfPower : fShadows[i]->fPower;
                     lRef->fD3DInfo.Diffuse.r
                         = lRef->fD3DInfo.Diffuse.g
                         = lRef->fD3DInfo.Diffuse.b
```

Other than that, it looks great 👍

File: Sources/Tests/.clang-tidy [hunk 8]
```diff
@@ -0,0 +1,2 @@
+InheritParentConfig: true
```
Comment (zrax) [inline Sources/Tests/.clang-tidy:1]:
```
[89/100] Running clang-tidy on test_CoreLib...
YAML:1:22: error: unknown key 'InheritParentConfig'
InheritParentConfig: true
                     ^~~~
```

Comment (Hoikas) [reply Sources/Tests/.clang-tidy:1]: It looks like this option is [relatively new](https://reviews.llvm.org/D75184).

Comment (zrax) [reply Sources/Tests/.clang-tidy:1]: That would explain it...  I'm using whatever version ships with MSVC

Comment (Hoikas) [reply Sources/Tests/.clang-tidy]: The best I can do is disable sanitizing the tests, unless you can write a regex to filter out `Sources/Tests/gtests` but not `Sources/Plasma` and `Sources/Tools`...

**Review by zrax — APPROVED** [review]

**Pull request merged**